### PR TITLE
[Parse] Remove 'ParseDeclOptions' parameter from 'parseDecl' function 

### PR DESCRIPTION
**Pull request by rintaro**
`ParseDeclOptions` can be trivially calculated from the current decl context. To reduce the number of the contextual parameters, calculate it inside the function.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by hamishknight — APPROVED** [review]
Nice!

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -5743,6 +5740,41 @@ setOriginalDeclarationForDifferentiableAttributes(DeclAttributes attrs,
...
+static Parser::ParseDeclOptions getParseDeclOptions(DeclContext *DC) {
+  using ParseDeclOptions = Parser::ParseDeclOptions;
+
+  if (DC->isModuleScopeContext())
+    return Parser::PD_AllowTopLevel;
+
+  if (DC->isLocalContext())
+    return Parser::PD_Default;
+
+  auto decl = DC->getAsDecl();
```
Comment (ahoppen) [inline lib/Parse/ParseDecl.cpp:5754]: Would it be worth checking if `decl != nullptr`? I don’t know when the parser might have a `DeclContext` that’s not a `Decl` (e.g. a closure) but seems like a potential source of crashes.

**Review by ahoppen — APPROVED** [review]
Nice

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (rintaro) [reply lib/Parse/ParseDecl.cpp:5754]: This _should_ always be non-null. All non-decl `DeclContext` cases should be handled above (i.e. local context, or mode scope context)

**Pull request merged**